### ISSUE TITLE
gobgp as an exabgp replacement

### ISSUE THREAD

**Issue by dbarrosop** [osrg/gobgp#1657]
Hello,
this is not really a bug. More like a suggestion for future enhancements so gobgp can be used as an orchestrator. The idea would be to have different neighbors on different tables so routes can be sent/received for each individual neighbor or group of neighbors. You can almost achieve this today using VRFs but they have extra logic to deal with the vpnv4 AF so it doesn't really work. So what I am suggesting is:

1. Have an "orchestrator" mode where you can create multiple tables an attach neighbors to them (similar to the VRF functionality but without extra logic)
2. This "orchestrator" mode would just insert routes into the different tables with the exact attributes the user specified, without any extra logic from gobgp, and send them the corresponding neighbors. This means it's up to the user to specify things like the path identifier, route-distinguisher, route-target communities, etc, if they are needed on the other end.

In summary, the end goal would be basically to be able to use gobgp as a replacement for exabgp.

Does this make sense? Happy to discuss.

Regards

**Comment by fujita** [comment]
"orchestrator" (controller, or such) is one of the major use cases for me so I'm interested in the enhancement for this direction. The vrf-neighbor feature doesn't work? You can play with VRFs without vpnv4 family.

**Comment by dbarrosop** [comment]
> The vrf-neighbor feature doesn't work? You can play with VRFs without vpnv4 family.

I might have made some mistake but when using VRFs I couldn't get gobgp to properly communicate with bird. The session would come up but as soon as I inserted the route into the table belonging to that VRF (gobgp side) the session would go down because of some missing field. Didn't investigate much but I figured it was because of the mismatch in the address families.

Another problem is that gobgp has logic in the VRF to set the RD and the necessary communities for the vpnv4 AFs, however, I wanted to set them myself and I couldn't either as there is some logic wired in gobgp to set them itself and to disallow users from doing it.

I might have missed something here, the point of this issue is to discuss how gobgp can be used for this purpose and I might be willing to help out in this direction as much as possible as I think gobgp would be a great replacement for my exabgp use cases (which fails to scale to hundreds of peers).

> You can play with VRFs without vpnv4 family.

That would be great but couldn't figure out how to do that with gobgp. Is it possible? How? Even in that mode I suspect there is logic around extended communities and maybe others that you may or may not want. Anyway, if you can point me in the right direction I might give it another try.

**Comment by fujita** [comment]
Here is how I configured.

I started gobgpd with the following minimum configuration. 
```
fujita@ubuntu:~/go/src/github.com/osrg/gobgp$ cat mini.conf 
[global.config]
as = 65001
router-id = "10.0.255.254"
```

I created a vrf and a peer with the vrf.
```
fujita@ubuntu:~$ gobgp vrf add vrf1 rd 10.100:100 rt both 10.100:100
fujita@ubuntu:~$ gobgp vrf
  Name                 RD                   Import RT            Export RT            ID   
  vrf1                 10.100:100           10.100:100           10.100:100           0    
fujita@ubuntu:~$ gobgp neighbor add 172.17.0.3 as 65003 vrf vrf1
```

The peer (bird) was up.
```
fujita@ubuntu:~/go/src/github.com/osrg/gobgp$ gobgp n
Peer          AS  Up/Down State       |#Received  Accepted
172.17.0.3 65003 00:01:22 Establ      |        0         0
```

I added some routes to the vrf1.
```
fujita@ubuntu:~$ gobgp vrf vrf1 rib add 10.0.0.0/24
fujita@ubuntu:~$ gobgp vrf vrf1 rib add 20.0.0.0/24
fujita@ubuntu:~$ gobgp vrf vrf1 rib
   Network              Next Hop             AS_PATH              Age        Attrs
*> 10.0.0.0/24          0.0.0.0                                   00:01:27   [{Origin: ?}]
*> 20.0.0.0/24          0.0.0.0                                   00:01:06   [{Origin: ?}]
```

The peer got the two routes.
```
bird> show route 
10.0.0.0/24        via 172.17.0.1 on eth0 [as65001 11:27:38] * (100) [AS65001?]
20.0.0.0/24        via 172.17.0.1 on eth0 [as65001 11:28:00] * (100) [AS65001?]
```

gobgp uses ipv4 family with the bird peer so doesn't add anything fancy to routes to be advertised.

https://github.com/osrg/gobgp/blob/master/table/path.go#L1179
https://github.com/osrg/gobgp/blob/master/table/path.go#L1226
The above functions are used receiving and sending routes with vrf-neighbors. When receiving, the former adds an extended community (rt) and when sending, the latter deleted the extended community. Looks like the latter deletes all the extended communities. I think that it's a bug. You should be able to send whatever attributes you want.

**Comment by dbarrosop** [comment]
> gobgp uses ipv4 family with the bird peer so doesn't add anything fancy to routes to be advertised.

Ok, I just realized the error. It was "on my end". So I was using the grpc interface to configure everything and when installing the paths I didn't specify the origin so bird was throwing me this error and shutting down the connection:

```
Error: Missing mandatory attribute: 01
```

That's probably a bug. Either the path shouldn't be installed at all or a default value should be defined. I will report it on a different issue.

What I am unable to make work now is add-path tx but that's surely a different story. I suspect the path-identifier is not being sent as I can see all the prefixes in the vrf rib but bird will only install the last one (I know bird is not the problem because it works just fine in production for me :) ). Will also report it on a different issue.

> Looks like the latter deletes all the extended communities. I think that it's a bug. You should be able to send whatever attributes you want

Yeah, that's kind of what I noticed.

Another related problem is that programmatic interface to install paths in gobgp is very cumbersome and hard to use. I'd love to be able to be able to send an object via grpc with every field set as a simple string or number (or arrays of strings and numbers) and have gobgp do all the necessary conversions on the server side.

**Comment by fujita** [comment]
Seems that there are more issues but the vrf-neighbor feature might work for your use-case?
>Another related problem is that programmatic interface to install paths in gobgp is very cumbersome and hard to use. 

Everyone says so, even I say so :)

https://github.com/osrg/gobgp/issues/1457

How about the above?

**Comment by dbarrosop** [comment]
> Seems that there are more issues but the vrf-neighbor feature might work for your use-case?

Yeah, I think it might work. I just need to make add-path work and then do some more testing.

> How about the above?

Ok, let me comment there about that.

Feel free to close this issue and maybe open a new one for the extended communities issue

**Comment by fujita** [comment]
I guess that all the related issues on this were fixed.

**Comment by dbarrosop** [comment]
thanks for all the awesome work :)
